Rapid Fire Game

teodorciuraru opened the pull request:
Rapid Fire game is a **quiz / trivia** game that lets the volunteers win points by answering questions, related to the application, under time pressure.

This PR includes a new module added to the application (the **Game** folder), new models (Game, RapidFireGame) and new screens (GamesViewController, AchievementsViewController (TODO), RapidFireViewController).

**TODO**: Save the users score persistently when the User Profile is created.

Commit:
Add first version of the quiz game.

bphenriques commented on malaria-ios/ViewControllers/Game.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+class Game: NSObject {
```
Is this working? It's because, if this is stored in the CoreData then I believe it is missing the @NSManaged tag before the instance variables like so:

```
@NSManaged var numberOfLevels: Int64
@NSManaged var name: String
```

You can generate the class from the model itself.

teodorciuraru replied:
Why should the games be stored in Core Data?

bphenriques replied:
Why did you inherit from NSObject ?

teodorciuraru replied:
Ah, by mistake. I removed it.

bphenriques commented on malaria-ios/ViewControllers/GamesViewController.swift:
```diff
@@ -0,0 +1,61 @@
+    view.backgroundColor = UIColor(patternImage: UIImage(named: "background")!)
+  }
+  
+  override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
```
Avoiding storyboard to refactor it afterwards?

bphenriques commented on malaria-ios/ViewControllers/GamesViewController.swift:
```diff
@@ -0,0 +1,61 @@
+extension GamesViewController: UICollectionViewDelegate {
+  
+  func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+    switch indexPath.row {
```
comment this part, it isn't clear what is happening

bphenriques commented on malaria-ios/ViewControllers/RapidFireGame.swift:
```diff
@@ -0,0 +1,29 @@
+    let entry2 = RapidFireGameEntry(question: "Question 2 (answer 1)", answers: ["Answer 1", "Answer 2", "Answer 3"], correctAnswer: 0)
+    let entry3 = RapidFireGameEntry(question: "Question 3 (answer 3)", answers: ["Answer 1", "Answer 2", "Answer 3"], correctAnswer: 2)
+    
+    entries.appendContentsOf([entry1, entry2, entry3])
```
Is it possible to import this questions from an JSON/Yaml/Xml/etc file? I think we have already some methods to import from JSON file (it would be a one step ahead if we later receive this from the server) ) ;)

Edit: regarding json, check sync manager

bphenriques commented on malaria-ios/ViewControllers/RapidFireGame.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    entries.appendContentsOf([entry1, entry2, entry3])
+    
+    super.init(numberOfLevels: entries.count, name: "Rapid Fire")
```
Move the variable a static field

teodorciuraru replied:
Can you please explain to me again?

bphenriques replied:
"Rapid Fire"

bphenriques commented on malaria-ios/ViewControllers/RapidFireGame.swift:
```diff
@@ -0,0 +1,29 @@
+  
+  let question: String, answers: [String], correctAnswer: Int
+  
+  init(question: String,
```
The init is short, put in the same line

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,115 @@
+  }
+  
+  func nextQuestion(correctAnswer: Bool) {
+    correctAnswer ? (userScore += 1) : showWrongAnswerAnimation()
```
Use if in this case.

I would recommend using only ternary operators on attributions or on method calls:
a = isTrue ? 0 : 1;
obj.CallMethod(isTrue ? 0 : 1)

Commit:
Add RapidFire Core Data Objects

And also other minor modifications: code refactoring, Rapid Fire Game's UI design etc.

bphenriques commented on malaria-ios/ViewControllers/AchievementsViewController.swift:
```diff
@@ -0,0 +1,16 @@
+    view.backgroundColor = UIColor(patternImage: UIImage(named: "background")!)
+  }
+  
+  override func didReceiveMemoryWarning() {
```
Remove this method. it just calls the parent. Having this method here doesn't add any more information.

bphenriques commented on malaria-ios/ViewControllers/GamesViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    // We switch VC's based on the index path that was selected
+    switch indexPath.row {
+    case 0: fallthrough
+    default: performSegueWithIdentifier("Show Rapid Fire", sender: self)
```
Move the Segue name to a constant (you are repeating the string twice as far as I see)

bphenriques commented on malaria-ios/ViewControllers/RapidFireGame.swift:
```diff
@@ -0,0 +1,28 @@
+
+class RapidFireGameEntry {
+  
+  let question: String, answers: [String], correctAnswer: Int
```
correctAnswer is a int? What means that variable?

Moreover, each variable to their own line:

``` swift
let x
let y
let z
```

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+      }
+      
+      // Set timer
+      timer = NSTimer.scheduledTimerWithTimeInterval(1, target: self, selector: #selector(updateTimer), userInfo: nil, repeats: true)
```
What is happening here? Move the 1 variable to an constant (the name of the contents should make things self-explanatory)

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+import UIKit
+
+class RapidFireViewController: UIViewController {
```
Tomorrow we will talk, since you simulating retrieving the questions from the server, what would happen if it returns 0 questions/answers, etc?

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+  
+  private var count = 5 {
+    didSet {
+      countDownLabel.text = count != 1 ? "\(count) seconds" : "\(count) second"
```
count + (count == 1 ? "second" : "seconds")

the ternary only handles the "seconds" issue. The == 1 is preferable to put positive conditions (I think you read code complete, there are mentions on this :) )

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+    view.backgroundColor = UIColor(patternImage: UIImage(named: "background")!)
+    
+    // Start game
+    currentLevel = 0
```
on viewDidLoad? Is it required to put this here instead of the class field (I will have to recall the life cycle of the view controllers). Or did you intended to put this on the viewWillAppear which makes more sense at this moment for me but I might not be aware of everything.

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+  }
+  
+  func showWrongAnswerAnimation() {
+    let defaultColor = questionLabel.textColor
```
Rename to originalColor. I think it is a more suitable name I think

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+  
+  var rapidFireGame: RapidFireGame?
+  
+  private var timer: NSTimer?
```
why optional?

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,159 @@
+  @IBInspectable let TimerBlinkAtSecond: Int = 3
+  @IBInspectable let TimerMaxValue: Int = 5
+  
+  var rapidFireGame: RapidFireGame?
```
why optional?

bphenriques commented on malaria-ios/NSManagedObjects/RestAPI/RFQuestion.swift:
```diff
@@ -0,0 +1,23 @@
+  @NSManaged public var title: String
+  @NSManaged public var correctAnswer: Int
+  
+  @NSManaged public var created_at: String
```
Why the extra fields?

bphenriques commented on malaria-ios/NSManagedObjects/RestAPI/CollectionRFQuestions.swift:
```diff
@@ -0,0 +1,12 @@
+
+public class CollectionRFQuestions: NSManagedObject {
+  
+  @NSManaged public var posts: NSSet
```
In my opinion, the questions shouldn't be retrieved as a "post"

Therefore they wouldn't have the same fields as the other endpoints (author,, created at, etc).

I would revise the NSManagedObject structure and simplify it to what is really required.

Commit:
Add several improvements and rule screen for games

bphenriques commented on malaria-ios/ViewControllers/RapidFireViewController.swift:
```diff
@@ -0,0 +1,206 @@
+  
+  private var GameProgressText: AlertText {
+    return ("Game over",
+            "You scored \(userScore / 3) "
```
Logic?

bphenriques commented on malaria-ios/ViewControllers/Constants.swift:
```diff
@@ -10,4 +10,8 @@ class Constants {
   static let localNotificationsArray = "Local Notifications Array"
+  
+  struct Games {
+    static let UnwindToGamesListVC = "Unwind to Games List"
```
I searched this, is this used?

teodorciuraru replied:
Forgot about that.

bphenriques commented on malaria-ios/ViewControllers/GameRulesViewController.swift:
```diff
@@ -0,0 +1,34 @@
+  }
+  
+  @IBAction func startGameBtnHandler(sender: AnyObject) {
+   performSegueWithIdentifier("Show " + game.name, sender: self)
```
Don't forget to add in the documentation how to extend the list of games! (the segue id detail is important for that)

Commit:
Add documentation regarding adding a new game to the list

Commit:
Add spacing between two buttons
